### ISSUE TITLE
Issue with building the Hexagon HVX DSP toolchain with TF

### ISSUE THREAD

**Issue by amirjamez** [tensorflow/tensorflow#18835]
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 14.04 
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.8r
- **Python version**:  2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**: 4.8.4
- **CUDA/cuDNN version**: 8
- **GPU model and memory**: GTX 1060
- **Exact command to reproduce**:

### Describe the problem

@satok16 @tensorflower-gardener , Could you please clarify the different version-combination of tools to use in an HVX-TF toolchain. It can save many people's effort to find and match these together. You have already mentioned in [HVX_TF](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx), that for `Qualcomm SDK 3.0`, the compatible nnlib version is [Aug-2017-commit](https://source.codeaurora.org/quic/hexagon_nn/nnlib/commit/?id=721b2d58f0f4e2d5b182f41e6b7c4db5356bf0fb). However, some features are missing there and we need to use the latest version with `SDK 3.3.3`, however, these libraries (`nnlib`, `libcontroller`, `TF-(sub)makefiles`) don't match, thus it has become a messy process to figure out which is related to what version. 

**1)** A sample error compiling libcontroller after compiling nnlib with latest SDK (3.3.3):

```
src_impl/hexagon_controller.c: In function 'hexagon_controller_AppendNode':
src_impl/hexagon_controller.c:484:70: error: 'hexagon_nn_output' has no member named 'max_size'
     pos += snprintf(&output_param_buf[pos], 500, "(%d), ", outputs[i].max_size);
                                                                      ^
make[1]: *** [android_Release/hexagon_controller.o] Error 1
make[1]: Leaving directory `~/Qualcomm/Hexagon_SDK/3.3.3/examples/common/hexagon_controller'
make: *** [tree] Error 2

```
**2)** If I use your suggested commit and SDK 3.0, this is the output I am having:

```
native : hexagon_graph_execution_test.cc:129 Read /data/local/tmp/img_299x299.bmp, size = 269156bytes
native : hexagon_graph_execution_test.cc:135 header size = 54
native : hexagon_graph_execution_test.cc:137 image size = 40
native : hexagon_graph_execution_test.cc:139 width = 299
native : hexagon_graph_execution_test.cc:141 height = -299
native : hexagon_graph_execution_test.cc:286 Ioading image finished.
native : hexagon_graph_execution_test.cc:185 Loading image finished.
native : hexagon_graph_execution_test.cc:189 Copy data to tensor.
native : hexagon_graph_execution_test.cc:307 Run graph
Init hexagon with max attributes (Controller version = 101)
native : hexagon_control_wrapper.cc:104 Add input: Mul, 0
native : hexagon_control_wrapper.cc:127 Allocate inout buffer
native : hexagon_control_wrapper.cc:304 Setup graph completed
Prepare failed! returned 0xffffffff

NN Id = -755923072
Execute graph!
Execution failed!
execute got err: -1

NN Id = -755923072
Execution failed
NN Id = -755923072
Failed to read data.
native : hexagon_graph_execution_test.cc:313 Output byte size = 4032
native : hexagon_graph_execution_test.cc:314 Output shape = [1,1008]
native : graph_transfer_utils.cc:47 === Dump ranking ===
native : graph_transfer_utils.cc:50 0: 1000, dumbbell, 0
native : graph_transfer_utils.cc:50 1: 999, carbonara, 0
native : graph_transfer_utils.cc:50 2: 998, stole, 0
native : graph_transfer_utils.cc:50 3: 997, rubber eraser, 0
native : graph_transfer_utils.cc:50 4: 996, coffee mug, 0
native : graph_transfer_utils.cc:50 5: 995, flagpole, 0
native : graph_transfer_utils.cc:50 6: 994, parallel bars, 0
native : graph_transfer_utils.cc:50 7: 993, cheeseburger, 0
native : graph_transfer_utils.cc:50 8: 992, bubble, 0
native : graph_transfer_utils.cc:50 9: 991, beaker, 0
Finalize hexagon
[       OK ] GraphTransferer.RunInceptionV3OnHexagonExampleWithTfRuntime (5848 ms)
[----------] 1 test from GraphTransferer (5848 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (5849 ms total)
[  PASSED  ] 1 test.

  YOU HAVE 5 DISABLED TESTS
```

**3)**  Also in the [toturial](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx), you did not mention anything about downloading/using an inception-v3 frozen-quantized model. If one follows the building each library from the source, where is the part related to use the [`tensorflow_inception_v3_stripped_optimized_quantized.pb`] ? Since, when I test with my custom quantized_frozen inception-v3 model, (renaming it to be the same as the original 2016 file (`tensorflow_inception_v3_stripped_optimized_quantized.pb`), I have the error output as:

```
...
native : hexagon_graph_execution_test.cc:533 Ioading image finished.
t1(loading image time)=0.026770
native : hexagon_graph_execution_test.cc:546 Build fused graph
native : remote_fused_graph_execute_utils.cc:259 Error during inference: Not found: FeedInputs: unable to find feed output Mul
native : graph_transfer_utils.cc:110 Check failed: status.ok()
Aborted

```
Thanks.

**Comment by karmel** [comment]
Tagging @satok16 for this one as well.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by karmel** [comment]
@satok16 , any comment on this?

**Comment by satok16** [comment]
@amirjamez
Which kind of hardware are you using?  Did you follow the troubleshoot in the tutorial?
amirjamez

**Comment by amirjamez** [comment]
@satok16 it is an open Q 820. Yes, I did. apparently, there is a mismatch between the lib-controller versions, nnlib, and the TF-toolchain. Libcontroller used in your toolchain (the one that downloaded the prebuilt library) is v90 and it belongs to a very old version of TF (maybe v1.1 or v1.2), thus using new TFs to build the library will result in failure. However, nnlib needs to be newer to support more ops and these initiates the conflicts.

**Comment by satok16** [comment]
"Prepare failed! returned 0xffffffff" means that it couldn't create NN in HVX.  So, I assume there may be some incompatibility issues on your hardware.  You may want to build nnlib by yourself for your hardware and that may work.  This feature is pretty much frozen mode, so you may want to wait for tflite for better hardware supports.

**Comment by amirjamez** [comment]
@satok16 thanks. I have built nnlib with `SDK 3.3.3`, but the test app provided in the test folder does not recognize the image when I pass. [Full info here.](https://stackoverflow.com/questions/50266601/running-the-nnlib-test-app-on-the-hexagon-hvx-dsp). Have you overcome this issue?

**Comment by satok16** [comment]
No, we don't use SDK 3.3 here as this is frozen mode.

**Comment by amirjamez** [comment]
@satok16, but what if someone wanted to use his modified `inception_v3` (newer than the one used in the toolchain as this was for December 2016) as input, it is needed to have the latest nnlib to support the ops and thus, one needs a newer version of SDK and lib controller to build those. So the whole chain will crumble again.

**Comment by satok16** [comment]
I see, if you want to get newer SDK version's support, I'd recommend you to wait for tflite's supports on nnlib as efforts are going there.

**Comment by amirjamez** [comment]
@satok16, could you at least help me out to pass the image mismatch issue that I am getting by executing the nnlib test app (graph_app). Looking at the code, no matter what parameters I pass, I am stuck with this issue:

`image size 25763 does not match element size 1, depth 3, width 299, height 299`

if I manually pass this segment (commenting out the check), I will have either small data, or, data overflow and lose adb connection everytime I execute it:

```
image size 25763 does not match element size 1, depth 3, width 92, height 92, elements 25763, area 8587
 filesize=25763 elementsize=1 height=92 width=92 depth=3
Run!
execute got err: -1
hexagon/ops/src/op_output.c:59:output 0 too small
Rank,Softmax,index,string
0,0.000000,1023,OVERFLOW!
1,0.000000,1,kit fox
2,0.000000,2,English setter
3,0.000000,3,Siberian husky
4,0.000000,4,Australian terrier
```

What is the correct `weight`, `width`, `depth` and `elements` parameters to be used with `keyboard_299x299.jpg`. Thanks.

**Comment by satok16** [comment]
Did you check the sample image which is used in build_and_run_inception_hexagon.sh?

**Comment by amirjamez** [comment]
@satok16, yes I did. Same issue [here](https://stackoverflow.com/questions/43995584/shell-gets-stuck-when-standalone-graph-app-in-hexagon-nnlib-is-ran) and [here](https://stackoverflow.com/questions/50266601/running-the-nnlib-test-app-on-the-hexagon-hvx-dsp). it is `img_299x299.bmp` and has 269156 bytes. Given the conditions in nnlib i need to have:

```
...
326         elements = filesize / elementsize;
327         area = elements / depth;
328         if (height * width == 0) {
329                 height = width = sqrt(area);
330                 assumed->height = height;
331                 assumed->width = width;
332         }
333         if  ((filesize % elementsize != 0)){
334                  ||(elements % depth != 0)
335                  (height * width != area) {
336                 printf("image size %d does not match "
...

```
thus, assuming a square box (width == height) and depth = 3 I need to have an image of size: 299*299*3 = 268203 which currently it is not (it is 269156)

**Comment by satok16** [comment]
"bitmap" has a header and you can't feed the bitmap binary to the inception v3 model as is.  You should load the image to a float array as hexagon_graph_execution_test.cc is doing.

**Comment by amirjamez** [comment]
@satok16 thanks. I see how you did it at the [hexagon_test.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/hexagon/hexagon_graph_execution_test.cc). So how about `.jpg` and `.dat`?  There is a keyboard_299x299.jpg in the test folder, doesn't it have to work out-of-the-box?

hmm, is that the reason there is a python script at `scripts/imagedump.py`. Does it get a .bmp and outputs just the .dat for that matter?

**Comment by satok16** [comment]
You can just convert the jpg to bitmap or directly to a float array in your program and feed it to the model.

**Comment by amirjamez** [comment]
Thanks, @satok16. Could you provide insights on why the bellow error occurs while executing the graph standalone app on HVX? 

```
Using </data/keyboard_299x299.dat>
filesize=268203 elementsize=1 height=299 width=299 depth=3
Run!
execute got err: -1
hexagon/ops/src/op_output.c:59:output 0 too small
output size=4096
Rank,Softmax,index,string
0,0.000000,1023,OVERFLOW!
1,0.000000,1,kit fox
2,0.000000,2,English setter
3,0.000000,3,Siberian husky
4,0.000000,4,Australian terrier
run failed: -1
```
This interestingly works with `SDK 3.0` and `nnlib august 2017`, but on both versions we have execution error -1.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by amirjamez** [comment]
@satok16 Could you let me know how to print all the append nodes in [Hexagon_control_wrapper.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/hexagon/hexagon_control_wrapper.cc) given inception-v3? I guess this way I may be able to debug the issue. Thanks

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by karmel** [comment]
@satok16 -- any update here?

**Comment by amirjamez** [comment]
@satok16 I managed to build the libraries and the hexagon_excution_graph adding `-E` to the [build toturial](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/hvx#build-tensorflow-linking-hexagon-library). It takes around 10 minutes to compile (after we already built lib_controller and lib_hexagon_skel.so) and I am searching to reduce its compilation time. 

Also, my other question was [unanswered here](https://github.com/tensorflow/tensorflow/issues/20206). I appreciate if you let me know how did you freeze and optimize the `inception-v3` model that doesn't have `tf.Squeeze()` inside. As you know, this method is not yet implemented on HVX. When I freeze and optimize for inference, these layers are still there, I guess some part of the graph should be stripped/removed but using the graph_transformation I was not able to reproduce the model. Thanks.
